01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+the build (space and version). This will be crucial for e.g. reporting the
+status of a pull request or commit back to GitHub.
+
+Add support for a "notifier" resource type. They have one hook, `notify`,
+which is invoked with various types of notifications
+(`build_started`, `build_finished`, as the poster children).
+
+
+# Caveats
+
+Terminology is now even more confusing. "Resource type" could mean either the
+"git" vs. "s3" or "notifier" vs. "artifact".
+
+
+# Open Questions
+
+## Can we reduce the `check` overhead?
+
+With spaces there will be more `check`s than ever. Right now, there's one
+container per recurring `check`. Can we reduce the container overhead here by
+requiring that resource `check`s be side-effect free and able to run in parallel?
+
+There may be substantial security implications for this.
+
+
+## Is `destroy` general enough to be a part of the interface?
+
+It may be the case that most resources cannot easily support `destroy`. One
+example is the `git` resource. It doesn't really make sense to `destroy` a
+commit. Even if it did (`push -f`?), it's a kind of weird workflow to support
+out of the box.
+
+Could we instead just have `put` and ensure that we `check` in such a way that
+deleted versions are automatically noticed? What would the overhead of this be?
+
+
+## Should `put` be given a space or return the space?
+
+The verb `PUT` in HTTP implies an idempotent action against a given resource. So
+it's intuitive that the `put` verb here would do the same.
+
+However, many of today's usage of `put` would be against a dynamically
+determined space. For example, most semver workflows involve `put`ing with the
+version determined by a file (often coming from the `semver` resource). So the
+space isn't known statically at pipeline configuration time.
+
+What's more, the resulting space for a semver push would only be `MAJOR.MINOR`,
+excluding the final patch segment. This is annoying to have to explicitly
+configure in your build.
+
+If we instead have `put` return both the space and the versions, this would be a
+lot simpler.
+
+
+## Should a notifier resource be able to "observe" another resource?
+
+This would allow e.g. a generic `git` resource and specific
+`github-commit-status` resource type rather than having to bake them all
+together.
+
+The downside here is that it would be introducing cross-resource-type interface
+contracts. Resource A would have to understand resource B's versions/spaces.
+This gets more possible with schema verification but still feels risky.
+
+
+# New Implications
+
+Here are a few use cases that resources were sometimes used for inappropriately:
+
+## Single-state resources
+
+Resources that really only have a "current state", such as deployments. This is
+still "change over time", but the difference is that old versions become invalid
+as soon as there's a new one.
+
+These can now be done by always marking the old versions as "deleted".
```

Comment:
how can a resource implementer make this happen for every single put, while simultaneously supporting explicit deletes from the pipeline side? I'd like first class support for these sorts of resources (especially because there are so many of them) but after reading this, and thinking about this exact problem myself prior to reading this, I'm still uncertain as to what that might look like.

Reply:
This hinges on `check` being able to detect that the version has gone away. If that's doable (big if, but I think it might be), then `put` can just be a general side-effecter, and `check` is the verification that things were either deleted (by noticing only the new version, and that the cursor version given to it is no longer there) or created (by having new versions).

Reply:
I know the resource that I maintain is simultaneously "single-state" and probably also spatial-- ie there's multiple underlying states, but each one invalidates itself when you update it. At any given time it's easy for me to list all the valid versions, but linearizing across spaces is cumbersome. 

When you say in an above comment
> When check is given a version that's no longer available, ...
If this is happening on a per-space basis, I think check not returning the cursor version makes enough sense-- things are pretty easy to linearize. My head swims a bit if think about deleting "old" versions or deleting versions from a big stack, though. Does "check" receive every version that's been emitted from `put`? does it receive every version that's been deleted?

Reply:
@cwlbraa Just circling back: I think this is now resolved. I've changed `put` to return two sets: a set of created versions, and a set of deleted versions, per-space. There was really no way `check` could magically discover deleted versions in all cases (only `git`). Tangentially related, I've changed `check` to return all versions in the event that given version is no longer found, per @itsdalmo's suggestion.